contrib/babelfishpg_tsql/src/procedures.c

```diff
@@ -2070,3 +2079,242 @@ gen_sp_droprolemember_subcmds(const char *user, const char *member)
+	{
+		if (((strlen(provider) == 7) && (strncmp(provider, "SQLNCLI", 7) == 0)) ||
+			((strlen(provider) == 10) && (strncmp(provider, "MSOLEDBSQL", 10) == 0)) ||
+			((strlen(provider) == 8) && (strncmp(provider, "SQLOLEDB", 8) == 0)))
+		{
+			/* if provider is a valid T-SQL provider, we throw a warning indicating internally, we will be using tds_fdw */
+			provider_warning = true;
+		}
+		else if ((strlen(provider) != 7) || (strncmp(provider, "tds_fdw", 7) != 0))
+			ereport(ERROR,
+				(errcode(ERRCODE_FDW_ERROR),
+				 	errmsg("Unsupported provider '%s'. Supported provider is 'tds_fdw'", provider)));
+
+		if (provstr != NULL)
+		{
+			/* we ignore provider string in any case */
+			provstr_warning = true;
+		}
+	}
+
+	ValidateLinkedServerDataSource(data_src);
+
+	stmt->servername = linked_server;
+	stmt->fdwname = "tds_fdw";
+	stmt->if_not_exists = false;
+
+	/* Add the relevant options */
+	options = lappend(options, makeDefElem("servername", (Node *) makeString(data_src), -1));
+
+	if (catalog != NULL)
+		options = lappend(options, makeDefElem("database", (Node *) makeString(catalog), -1));
+
+	stmt->options = options;
+
+	CreateForeignServer(stmt);
+
+	/* We throw warnings only if foreign server object creation succeeds */
+	if (provider_warning)
+	{
+		char *msg = "Warning: Using the TDS Foreign data wrapper (tds_fdw) as provider";
+
+		ereport(WARNING, errmsg("%s", msg));
+
+		if (*pltsql_protocol_plugin_ptr && (*pltsql_protocol_plugin_ptr)->send_info)
+			((*pltsql_protocol_plugin_ptr)->send_info) (0, 1, 0, msg, 0);
+	}
+
+	if (provstr_warning)
+	{
+		char *msg = "Warning: Ignoring @provstr argument value";
+
+		ereport(WARNING, errmsg("%s", msg));
+
+		if (*pltsql_protocol_plugin_ptr && (*pltsql_protocol_plugin_ptr)->send_info)
+			((*pltsql_protocol_plugin_ptr)->send_info) (0, 1, 0, msg, 0);
+	}
+
+	return (Datum) 0;
+}
+
+Datum
+sp_addlinkedsrvlogin_internal(PG_FUNCTION_ARGS)
+{
+	char *servername = PG_ARGISNULL(0) ? NULL : text_to_cstring(PG_GETARG_TEXT_P(0));
+	char *useself = PG_ARGISNULL(1) ? NULL : text_to_cstring(PG_GETARG_TEXT_P(1));
+	char *username = PG_ARGISNULL(3) ? NULL : text_to_cstring(PG_GETARG_TEXT_P(3));
+	char *password = PG_ARGISNULL(4) ? NULL : text_to_cstring(PG_GETARG_TEXT_P(4));
+
+	CreateUserMappingStmt *stmt = makeNode(CreateUserMappingStmt);
+	RoleSpec *user = makeNode(RoleSpec);
+	List *options = NIL;
+
+	stmt->servername = servername;
+	stmt->if_not_exists = false;
+
+	user->roletype = ROLESPEC_CURRENT_USER;
+	user->location = -1;
+	stmt->user = user;
+
+	/* We do not support login using user's self credentials */
+	if ((useself == NULL) || (strlen(useself) != 5) || (strncmp(downcase_identifier(useself, 5, false, false), "false", 5) != 0))
+		elog(ERROR, "Only @useself = FALSE is supported");
+
+	/* Add the relevant options */
+	options = lappend(options, makeDefElem("username", (Node *) makeString(username), -1));
+	options = lappend(options, makeDefElem("password", (Node *) makeString(password), -1));
```

Comment:
if we are expecting that there would always be these two options then better to use list_make2()